### PR TITLE
Fix reconcileDelete during namespace delete

### PR DESCRIPTION
**Pull request by gibizer**
`oc delete namespace openstack` hangs forever because KeystoneAPI is reconciled in the state when the namespace already being terminated but the KeystoneAPI CR has no delete timestamp set yet. In this state KeystoneAPI's reconciler tries to manipulate things in the namespace (service account, TransportURL, etc) and that is rejected by k8s. This cause KeystoneAPI to become Ready=False. This then prevent KeystoneEndpoint and KeystoneService deletion to proceed as they are waiting for KeystoneAPI to become Ready=true.

This patch implements one possible solution. It assumes if the KeystoneAPI is being deleted then there is no point removing the endpoints and users from the keystone DB as the DB will be deleted too. Therefore the KeystoneEndpoint and KeystoneService simply removes the finalizers if they detect that KeystoneAPI deleted timestamp is set.

This partially resolves the namespace deletion hangs. (There is another issue with DNSMasq CR that still blocks the namespace deletion)

Fixes: [OSP-27793](https://issues.redhat.com//browse/OSP-27793)

### CODE REVIEW

File: controllers/keystoneservice_controller.go [hunk 1]
```diff
@@ -170,6 +170,15 @@ func (r *KeystoneServiceReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 		return ctrl.Result{}, err
 	}
 
+	// If both the endpoint and the KeystoneAPI is deleted then we can skip
+	// the cleanup of the endpoint in the DB as the DB is going away as well.
+	// Moreover if KeystoneAPI is being deleted then we cannot talk to the
+	// keystone REST API any more. This happens for example during namespace
+	// deletion.
```
Comment (abays) [inline controllers/keystoneservice_controller.go:177]: Maybe the comment here should speak of `service` instead of `endpoint`?

Comment (gibizer) [reply controllers/keystoneservice_controller.go:177]: yepp. Copy-paste :(

**Comment by gibizer** [comment]
Together with https://github.com/openstack-k8s-operators/infra-operator/pull/178 this fixes the hanging namespace deletion

**Review by abays — APPROVED** [review]
/lgtm

**Pull request merged**